### ISSUE TITLE
(Optional/pragma) Allow records to be used as the 2nd argument to `setmetatable`, and automatically set `.__index`

### ISSUE THREAD

**Issue by Frityet** [teal-language/tl#786]
If you want to use metamethod on instances on records, you have to do some hackery right now

```lua
local record Person
    name: string
    age: integer

    metamethod __add: function(Person, Person): Person
end
rawset(Person, "__index", Person)

function Person:create(name: string, age: integer): Person
    return setmetatable({
        name = name,
        age = age,
    }, Person as metatable<Person>)
end

function Person:say_hello()
    print("Hello, my name is "..self.name)
end

function Person:__add(other: Person): Person
    return Person:create("Child of "..self.name.." and "..other.name, 0)
end

function Person:__tostring(): string
    return self.name.." (age "..self.age..")"
end

local p1 = Person:create("Alice", 20)
local p2 = Person:create("Bob", 30)

local p3 = p1 + p2

print(p1)
p1:say_hello()

print(p2)
p2:say_hello()

print(p3)
p3:say_hello()

```

An idea to remedy this is to add a pragma that 
1. sets the `__index` metafield to the record
2. Allows the record to be used in place of a metamethod

```lua
--#pragma class
local record Person
    name: string
    age: integer

    metamethod __add: function(Person, Person): Person
end

function Person:create(name: string, age: integer): Person
    return setmetatable({
        name = name,
        age = age,
    }, Person)
end

function Person:say_hello()
    print("Hello, my name is "..self.name)
end

function Person:__add(other: Person): Person
    return Person:create("Child of "..self.name.." and "..other.name, 0)
end

function Person:__tostring(): string
    return self.name.." (age "..self.age..")"
end

local p1 = Person:create("Alice", 20)
local p2 = Person:create("Bob", 30)

local p3 = p1 + p2

print(p1)
p1:say_hello()

print(p2)
p2:say_hello()

print(p3)
p3:say_hello()
```

Another feature that could be nice is for the record itself to have it's `__name` metafield set to improve output, and can be used to identify the record and instances:

```lua
--#pragma class with_name
local record Point
    x: integer
    y: integer    
end
```

turns into

```lua
local Point = {}
Point.__index = "Point"
Point.__name = "Point"
setmetatable(Point, { __name = "record Point" })

print(Point) --on lua 5.4 would print record Point: 0xDEADBEEF
```

**Comment by hishamhm** [comment]
That would be effectively blessing one particular object models as a built-in one. That's not something I'm planning to do at this time. A generalized macro system (or other form compile-time reflection) could allow for one to implement this feature for their own object model. But in a mechanisms-not-policies spirit of Lua, I don't want Teal to ship with support for one specific object model, causing other existing Lua frameworks to feel second-class.

Also, the pragma proposal is not intended to double as annotations (the kinds of things you usually see with `@-` syntax in other languages). Pragmas are intended as compiler directives, more akin to compiler flags (e.g. the `#pragma` use in C compilers). I understand these sit in a gray area, but I don't think there's a well-established definition in the literature.

My practical goal for pragmas is to allow for handling compatibility issues when dealing with the language evolution. That is, in a Teal codebase with no legacy concerns, there should be no pragmas.

So, in short, setting `__index` and casting the metatable for this particular model exemplified here will still be needed. But yes, long term I'd like to have a solution to reduce OOP boilerplate.

**Comment by hishamhm** [comment]
(I always forget that Github's Close button has this "completed vs not planned" binary; I don't like it very much, so please ignore that -- the comment itself has the proper nuance.)
